### ISSUE TITLE
Partially Wrapped Array Messaging

### ISSUE THREAD

**Issue by ptheywood** [FLAMEGPU/FLAMEGPU2#593]
As of #549, wrapped array messages are available via `.wrap` for the 1, 2 and 3 dimensional array message types.

This wraps in all dimensions of the message, which is probably desirable in some cases. 

In other cases, only wrapping on 1 or 2 dimensions may be more appropraite. I.e. a 3D environment that wraps in all 3 Dimensions can't be visualised using 3D geometry. 
Instead, prioviding wrapping in 1 or 2 dimensions for a 3D system provides cylindrical and torroidal geometries. 

+ [ ] 2D Array messaging with wrapping in 1 dimension
+ [ ] 3D Array messaging with wrapping in 1 dimension
+ [ ] 3D Array messaging with wrapping in 2 dimensions

To minimise API changes, `.wrap` could always provide full wrapping, while higher demensions will provide additional methods `.wrap1D` and `.wrap2D` or similar where appropraite?

Alternatively it could potentially be specified at model description time? by allowing users to set whether wrapping should occur in each dimension? This might be difficult / costly to implement in practice however.

**Comment by Robadob** [comment]
> Alternatively it could potentially be specified at model description time? by allowing users to set whether wrapping should occur in each dimension? This might be difficult / costly to implement in practice however.

Unless it becomes a different messaging type it will incur extra branches, hence the proposal for a distinct wrap method for each weird wrap format.

**Comment by ptheywood** [comment]
> Unless it becomes a different messaging type it will incur extra branches (inside the device code), hence the proposal for a distinct wrap method for each weird wrap format.

This is why i mentioned it would potentially be coslty. I think there are template-based solutions to achieve this, but with swig etc that might present issues, unsure without trying.

**Comment by Robadob** [comment]
I don't think it would be possible for it to be specified at model desc time, short of it being a different message type. The template dependency, for the value specified at desc time, to reach the agent function (which might be in a different compilation unit) isn't very realistic.
